src/slice.rs

```diff
@@ -48,6 +48,48 @@ impl Slice {
     pub fn step_by(self, step: Ixs) -> Self {
         Slice { step, ..self }
     }
+
+    /// Returns the `n`th index in this slice.
+    ///
+    /// The `n` argument starts from zero, so `nth(0)` is the first index. If
+    /// `n` is greater than or equal to the length of the slice, then the
+    /// return value is `None`.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use ndarray::Slice;
+    ///
+    /// // indices 0, 1, 2, 3, 4
+    /// assert_eq!(Slice::new(0, Some(5), 1).nth(2), Some(2));
+    ///
+    /// // indices 1, 3
+    /// assert_eq!(Slice::new(1, Some(5), 2).nth(1), Some(3));
+    ///
+    /// // indices 1, -1, -3
+    /// assert_eq!(Slice::new(1, Some(-4), -2).nth(2), Some(-3));
+    ///
+    /// // indices 1, 3, 5, ...
+    /// assert_eq!(Slice::new(1, None, 2).nth(2), Some(5));
+    ///
+    /// // indices 3, 3, 3, ...
+    /// assert_eq!(Slice::new(3, Some(5), 0).nth(1000), Some(3));
```

Comment:
What meaning do we give to a step of 0? It used to be an error when using it for slicing, I'll have to check again.

Reply:
If you actually call `slice()` with a step of 0, then you'll get a panic. (See [`dimension::do_slice`](https://github.com/bluss/rust-ndarray/blob/3d9134bdf5b35c0a4f70e04590f5e0a4cc6d07c6/src/dimension/mod.rs#L264).) You'll also get a panic when calling `slice()` if the slice begin/end is out-of-range.

In contrast, `nth()` doesn't panic in these cases. It's possible to change it to panic if `step == 0`, without changing the function signature. To make it panic for out-of-range indices, it would need an argument for the length of the axis you're going to slice.

If you'd prefer it panic in those cases, I'd probably choose a different name (e.g. `index()`) instead of `nth()`.

Reply:
Great, thanks for clarifying!

That sounds you have no particular behaviour in mind for 0, so we can choose one freely. I'm not too fond of mixing panics in explicitly checked or fallible methods either, so I'm not sure.  I do lean towards that it should be a panic -- maybe a debug assertion. Maybe indeed a debug assertion already in Slice::new ? Or is that redundant?

Reply:
Ah with public fields a debug assertion in new doesn't do much good anyway.

Reply:
It's certainly possible to make the fields non-public (or maybe just `step`) and add an assertion to `new()` and `step_by()`. In some ways, I think that's a better design anyway; what do you think? It's still not possible to handle out-of-range indices this way, but it would catch `step == 0` early.

If we make some of the fields in `Slice` non-public, I would also change `enum SliceOrIndex { Slice { start, end, step }, ... }` to `enum SliceOrIndex { Slice(Slice), ... }` to hide the fields there too.